server/audit_math/suite.py

```diff
@@ -0,0 +1,660 @@
+            raise ValueError("Null is impossible, given the sample")
+
+        if lower_n_w_limit > upper_n_w_limit:
+            lower_n_w_limit, upper_n_w_limit = upper_n_w_limit, lower_n_w_limit
+
+        LR_derivative = (
+            lambda Nw: np.sum([1 / (Nw - i) for i in range(n_w)])
+            + np.sum([1 / (Nw - null_margin - i) for i in range(n_l)])
+            - 2
+            * np.sum(
+                [1 / (self.num_ballots - 2 * Nw + null_margin - i) for i in range(n_u)]
+            )
+        )
+
+        # Sometimes the upper_n_w_limit is too extreme, causing illegal 0s.
+        # Check and change the limit when that occurs.
+        if np.isinf(null_logLR(upper_n_w_limit)) or np.isinf(
+            LR_derivative(upper_n_w_limit)
+        ):
+            upper_n_w_limit -= 1
+
+        # Check if the maximum occurs at an endpoint: deriv has no sign change
+        if LR_derivative(upper_n_w_limit) * LR_derivative(lower_n_w_limit) > 0:
+            nuisance_param = (
+                upper_n_w_limit
+                if null_logLR(upper_n_w_limit) >= null_logLR(lower_n_w_limit)
+                else lower_n_w_limit
+            )
+        # Otherwise, find the (unique) root of the derivative of the log likelihood ratio
+        else:
+            nuisance_param = sp.optimize.brentq(
+                LR_derivative, lower_n_w_limit, upper_n_w_limit
+            )
+        logLR = alt_logLR - null_logLR(nuisance_param)
+        LR = float(np.exp(logLR))  # This value is always a float, but np.exp
+        # can return a vector. casting for the typechecker.
+
+        return 1.0 / LR if 1.0 / LR < 1 else 1.0
+
+
+class BallotComparisonStratum:
+    """
+    A class encapsulating a stratum of ballots in an election. Each stratum is its
+    own contest object, with its own margin. Strata, along with the overall
+    contest object, are passed to the SUITE module when perfoming mixed-strata
+    audits.
+    """
+
+    num_ballots: int
+    vote_totals: Dict[str, int]
+    misstatements: Dict[Tuple[str, str], Dict[str, int]]
+    results: CVRS
+    sample_size: int
+
+    def __init__(
+        self,
+        num_ballots: int,
+        vote_totals: Dict[str, int],
+        results: CVRS,
+        misstatements: Dict[Tuple[str, str], Dict[str, int]],
+        sample_size: int,
+    ):
+        """
+        Initializes the ballot comparison stratum.
+
+        Inputs:
+            contest: contest information for this stratum, including candidate
+                     vote totals and stratum size
+            results: The CVRs for this stratum.
+            misstatements: All of the misstatements observed so far, by winner-loser pair. E.g.,
+                    {
+                        (winner1, loser1): {"o1": 0, "o2": 0, "u1": 0, "u2": 0},
+                        (winner1, loser2): ...,
+                        ...,
+                        (winner2, loser1): ...
+                    }
```

Comment:
where do you get winner-loser pairs from if you are creating a BallotComparisonStratum?

Reply:
(winner, loser) come from the overall contest results. They aren't the winners/losers of the stratum, but of the overall contest.